cores/arduino/stm32/dma.c

```diff
@@ -0,0 +1,293 @@
+DISCLAIMED. IN NO EVENT SHALL WI6LABS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  ******************************************************************************
+  */
+
+#include "stm32_def.h"
+#include "dma.h"
+#include "analog.h"
+
+#ifdef __cplusplus
+ extern "C" {
+#endif
+
+extern ADC_HandleTypeDef AdcHandleTab[ADC_NUM];
+
+static bool dmaTab[DMA_NUMBER][CHANNEL_NUMBER];
+
+/**
+  * @brief  This function will init the dma and channel corresponding 
+  *         to the wanted peripheral
+  * @param  periph_handle : pointer to the peripheral HandleTyeDef
+  * @param  periph        : the peripheral type to use
+  * @retval none
+  */
+void dma_allocate(void *periph_handle, Peripheral periph, void **channel, IRQn_Type *irqHandler, uint8_t *indice)
+{
+  int8_t    channelNumber = 0;
+  int8_t    dmaNumber     = 0;
+
+  ADC_HandleTypeDef *hadc = NULL;
+  //SPI_HandleTypeDef *spi  = NULL;  // to do
+  
+  switch(periph) {
+    case ADC_PERIPH :
+    {
+      hadc = (ADC_HandleTypeDef*)periph_handle;
+      *channel = pinmap_find_dmaChannel(hadc->Instance, DmaMap_DMA);
+      
+      /* Set clock as active */
+      channelNumber = pinmap_find_dmaChannelNumber(hadc->Instance, DmaMap_DMA);
+      dmaNumber     = pinmap_find_dma(hadc->Instance, DmaMap_DMA);
+      
+      /* NVIC configuration for DMA Input data interrupt */
+      *irqHandler = pinmap_find_dmaIrqHandler(hadc->Instance, DmaMap_DMA);
+      #ifdef ADC1_BASE
+      if (hadc->Instance == ADC1)
+      {
+        *indice = 0;
+      }
+      #endif
+      #ifdef ADC2_BASE
+      if (hadc->Instance == ADC2)
+      {
+         *indice = 1;
+      }
+      #endif
+      #ifdef ADC3_BASE
+      if (hadc->Instance == ADC3)
+      {
+        *indice = 2;
+      }
+      #endif
+      #ifdef ADC4_BASE
+      if (hadc->Instance == ADC4)
+      {
+        *indice = 3;
+      }
+      #endif
+    };break;
+    /*case SPI :
+    {
+      spi = (SPI_HandleTypeDef*)periph_handle;
+    };break;*/ //to do
+  }
+  
+#ifdef __HAL_RCC_DMA1_CLK_ENABLE
+  __HAL_RCC_DMA1_CLK_ENABLE();
+#endif
+
+#ifdef __HAL_RCC_DMA2_CLK_ENABLE
+  __HAL_RCC_DMA2_CLK_ENABLE();
+#endif
+  
+  /* Set clock as active */
+  dmaTab[dmaNumber-1][channelNumber] = true;
+}
+
+/**
+  * @brief  This function will deinit the dma and channel corresponding 
+  *         to the wanted peripheral
+  * @param  periph_handle : pointer to the peripheral HandleTyeDef
+  * @param  periph        : the peripheral type to use
+  * @retval none
+  */
+void dma_free(void *periph_handle, Peripheral periph)
+{
```

Comment:
same as allocate, simply pass the Instance instead of handle and periph type.
Or peripheral instance and DMA handle if handle is needed.
In this way you don't need ADC types

Reply:
For dma_free, i try to remove the peripheral type, but I've got a problem for that call : 
HAL_DMA_DeInit(hadc->DMA_Handle);
I can't DeInit the DMA. When i think about it, it's not a big deal, because we stop peripheral interruption by doing this : 
HAL_NVIC_DisableIRQ(irqHandler);
So this peripheral will not generate other interruption, and the DMA will not have other wok to do but it is always initialize. 
What do think about that? Is it ok for you?

Reply:
I think that you should pass dma_free(Instance, DmaHandle) as both are known at driver level.